### DISCUSSION TITLE
Questions about the highlighting system

### DISCUSSION THREAD

**Discussion by molleweide** [ggandor/lightspeed.nvim#89] [Q&A]
Hi,

If I search for a char that there are many instances of some hints light up in `white` and some `red` and usually the red chars are all the same. What is the purpose of these. I have read the usage guide a couple of time but this is stilla mystery to me.
If I press the red char the cursor moves but I don't understand the pattern of where it is moved.

Also sometimes if i search for `f` then the white hints show in one place `fa` and then in another place `f` alone and so I don't understand how to navigate to the lone f.

**Comment by molleweide** [comment]
hold on I need to read the usage guide a couple more times.

**Reply by molleweide** [reply]
i guess one big thing it boils down to is that I don't understand how to access the red chars without red background. but I am not sure lol.

**Comment by ggandor** [comment]
It just occurred to me, after writing all this stuff below, that there is a cool video available now, explaining the usage: https://www.youtube.com/watch?v=ESyld9NCl1w

Let's say your target is `ab`.

* After pressing `s` or `S`, the unique characters are highlighted (white). If there is only one `a` in the whole window, then it will be highlighted, and you will jump just by pressing `a`.

* After entering the first char of the search pattern (i.e., you have typed `sa` or `Sa` so far), you see _all_ bigrams starting with `a` (`ax`, `ay`, `az`, ...) highlighted in some manner. 
    * Some matches are just highlighted with white foreground, just like the unique chars in the previous step - if you enter `x`, `y`, etc. now, i.e., the second char, you will jump. 
    * If the label has a red background, you can just press the label, and you will jump too (you do not _have to_ though). Those are the so called "shortcuts".
    * Otherwise, if you see a red-underlined label, without a background, you need to _continue_ with the pattern, that is, entering `b`.

* After entering `b`, only the `ab` matches remain highlighted, and you have already jumped to the _first_ of the `ab` matches automatically. You can continue editing or enter one of the labels now.

**Comment by molleweide** [comment]
Hey man!! Thanks for your write up to me. I think that I might be starting to understand now. This is quite amazing when the muscle memory and reflexes really are starting to kick in. Holy shit!!!



One thing that I have noticed is this and I would like your comment on it. This pertains to red underlined chars.

1. Target is `ab` in word `abcdef`
2. Sometimes when I want to jump to target it puts a red underlined char, eg. `u` on `b`.
3. In order to jump to ab I then need to remember what char was under `u`.
4. Is this suppose to happen? Do you understand what I mean?

**Reply by ggandor** [reply]
That implies that `cd` is also a match, so that's why the label should be displayed over `b`. However, in that case, you will see `b` displayed over `a` (with pale brownish color in the default scheme). The only exception is when `ab` itself is overlapped by the label of another match _before_ - it's a very rare case, but then `b` will not be visible at all, yeah.
